src/services/services.ts

```diff
@@ -4316,6 +4361,159 @@ namespace ts {
+
+
+                if (directoryProbablyExists(baseDir, host)) {
+                    // Enumerate the available files
+                    const files = host.readDirectory(baseDir, extensions, /*exclude*/undefined, /*include*/["./*"]);
+                    forEach(files, (f) => {
+                        const fName = includeExtensions ? getBaseFileName(f) : removeFileExtension(getBaseFileName(f));
+
+                        if (startsWith(fName, toComplete)) {
+                            result.push({
+                                name: fName,
+                                kind: ScriptElementKind.unknown,
+                                kindModifiers: ScriptElementKindModifier.none,
+                                sortText: fName
+                            });
+                        }
+                    });
+
+                    // If possible, get folder completion as well
+                    if (host.getDirectories) {
+                        const directories = host.getDirectories(baseDir);
+                        forEach(directories, (d) => {
+                            const dName = getBaseFileName(removeTrailingDirectorySeparator(d));
+
+                            if (startsWith(dName, toComplete)) {
+                                result.push({
+                                    name: ensureTrailingDirectorySeparator(dName),
+                                    kind: ScriptElementKind.unknown,
+                                    kindModifiers: ScriptElementKindModifier.none,
+                                    sortText: dName
+                                });
+                            }
+                        });
+                    }
+                }
+
+                return includeExtensions ? result : deduplicate(result, (a, b) => a.name === b.name);
+            }
+
+            /**
+             * Check all of the declared modules and those in node modules. Possible sources of modules:
+             *      Modules that are found by the type checker
+             *      Modules from node_modules (i.e. those listed in package.json)
+             *          This includes all files that are found in node_modules/moduleName/ with acceptable file extensions
+             */
+            function getCompletionEntriesForNonRelativeModules(fragment: string, scriptPath: string): CompletionEntry[] {
+                return ts.map(enumeratePotentialNonRelativeModules(fragment, scriptPath), (moduleName) => {
+                    return {
+                        name: moduleName,
+                        kind: ScriptElementKind.unknown,
+                        kindModifiers: ScriptElementKindModifier.none,
+                        sortText: moduleName
+                    };
+                });
+            }
+
+            function enumeratePotentialNonRelativeModules(fragment: string, scriptPath: string) {
+                const ambientSymbolNameRegex = /^"(.+?)"$/;

```

Comment:
lift this outside the function to avoid creating a new object every time the function is called.